2-planning/2-mocking.adoc

```diff
@@ -0,0 +1,137 @@
+etc.), or you can use https://my-json-server.typicode.com/[My JSON Server] - the
+hosted version from the same author.
+
+image::images/mocks-my-json-server.png[]
+
+Now, at this point you are probably thinking: "That's lovely and all, but
+maintaining that one JSON file is going to be a freaking nightmare!"
+
+Yup. Adding a single property to a resource means you are going to have to go
+through each record in `db.json`. Beyond that, formatting the file can turn into
+a pain in the ass, conflicts will be rife as the file grows and different
+developers add different properties, etc. It's good for quick and dirty demos,
+but is probably not something you will want to use in anger.
+
+How about instead of manually maintaining `db.json`, a tool could be pointed at API contracts, and a mock server was automagically created?! Then it figure out its own
+examples based on the examples/default values supplied in the contract, and if the description language has validation rules it could even apply those too!
+
+Thankfully loads of folks in the API ecosystem thought about this, and there is a wide selection of tooling.
+
+== Hosted Editors Love Mock Servers
+
+Most of the hosted editors covered so far have mock servers built in, and often it is either a case of enabling it, or it is already enabled and you just need to find the URL to the server instance.
+
+== Standard HTTP Mocking
+
+=== Hosted Mock Servers
+
+https://swagger.io/tools/swaggerhub/[SwaggerHub]
+
+http://stoplight.io/[Stoplight]
+
+// TODO Maybe I'll add https://getsandbox.com/ but yesterday on a github issue they thanks somebody for reminding them OpenAPI v3.0 existed and its 2019!
+
+// TODO creators of wiremock have a thing http://get.mocklab.io/ - import "swagger" and integrates with swaggerhub
+
+=== Free/Open-Source Tools
+
+http://stoplight.io/platform/prism/[Prism]
+
+https://github.com/danielgtaylor/apisprout[API Sprout]
+
+As always check https://openapi.tools/[OpenAPI.Tools] for the latest offering of mocking tools.
+
+// TODO Can JSON Schema do a decent mock?
+// https://www.npmjs.com/package/mock-json-schema
+// https://www.npmjs.com/package/cast-with-schema
+
+== GraphQL Mocking
+
+https://app.graphqleditor.com/[GraphQL Editor Cloud]
+
+https://www.apollographql.com/docs/graphql-tools/mocking.html[Apollo - GraphQL Tools]
+
+== gRPC Mocking
+
+gRPC has two mocking tools which require a bit more setup. They seem to be more useful for unit/integration testing (which will be covered in later sections) than helping at a planning stage. We will look at testing later on.
+
+// TODO can these be used for mock servers like the above? https://github.com/YoshiyukiKato/grpc-mock
+// https://github.com/jekiapp/gripmock
+
+== Interacting with this Mock
+
+Interacting with a mock server should be incredibly similar to how you would interact with any API. That is the point, after-all.
+
+Just find the instance of the mock server, then point your HTTP/GraphQL/gRPC client at it.
```

Comment:
```suggestion
Find the instance of the mock server, then point your HTTP/GraphQL/gRPC client at it.
```

Reply:
I'm a fan of avoiding using words like "Just", "Simply" etc when explaining things because it may not be that obvious to the other person, and this just makes them feel like they're an idiot for not knowing things. Obviously ignore me if you disagree! 😇